diff --git llvm/include/llvm/MC/MCSubtargetInfo.h llvm~patched/include/llvm/MC/MCSubtargetInfo.h
index 09130c464..1d9c98220 100644
--- llvm/include/llvm/MC/MCSubtargetInfo.h
+++ llvm~patched/include/llvm/MC/MCSubtargetInfo.h
@@ -222,6 +222,10 @@ public:
     return Found != ProcDesc.end() && StringRef(Found->Key) == CPU;
   }

+  ArrayRef<SubtargetSubTypeKV> getCPUTable() const { return ProcDesc; }
+
+  ArrayRef<SubtargetFeatureKV> getFeatureTable() const { return ProcFeatures; }
+
   virtual unsigned getHwMode() const { return 0; }

   /// Return the cache size in bytes for the given level of cache.
